engines/sherlock/journal.cpp

```diff
@@ -0,0 +1,1212 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "sherlock/journal.h"
+#include "sherlock/sherlock.h"
+
+namespace Sherlock {
+
+#define JOURNAL_BUTTONS_Y 178
+#define LINES_PER_PAGE 11
+
+// Positioning of buttons in the journal view
+const int JOURNAL_POINTS[9][3] = {
+	{ 6, 68, 37 },
+	{ 69, 131, 100 },
+	{ 132, 192, 162 },
+	{ 193, 250, 221 },
+	{ 251, 313, 281 },
+	{ 6, 82, 44 },
+	{ 83, 159, 121 },
+	{ 160, 236, 198 },
+	{ 237, 313, 275 }
+};
+
+const int SEARCH_POINTS[3][3] = {
+	{ 51, 123, 86 },
+	{ 124, 196, 159 },
+	{ 197, 269, 232 }
+};
+
+/*----------------------------------------------------------------*/
+
+Journal::Journal(SherlockEngine *vm): _vm(vm) {
+	// Initialize fields
+	_maxPage = 0;
+	_index = 0;
+	_sub = 0;
+	_up = _down = false;
+	_page = 1;
+	_converseNum = -1;
+
+	// Load the journal directory and location names
+	loadJournalLocations();
+}
+
+/**
+ * Records statements that are said, in the order which they are said. The player
+ * can then read the journal to review them
+ */
+void Journal::record(int converseNum, int statementNum, bool replyOnly) {
+	int saveIndex = _index;
+	int saveSub = _sub;
+
+	// Record the entry into the list
+	_journal.push_back(JournalEntry(converseNum, statementNum, replyOnly));
+	_index = _journal.size() - 1;
+
+	// Load the text for the new entry to get the number of lines it will have
+	int newLines = loadJournalFile(true);
+
+	// Restore old state
+	_index = saveIndex;
+	_sub = saveSub;
+
+	// If new lines were added to the ournal, update the total number of lines
+	// the journal continues
+	if (newLines) {
+		_maxPage += newLines;
+	} else {
+		// No lines in entry, so remove the new entry from the journal
+		_journal.remove_at(_journal.size() - 1);
+	}
+}
+
+/**
+ * Load the list of location names that the journal will make reference to
+ */
+void Journal::loadJournalLocations() {
+	Resources &res = *_vm->_res;
+
+	_directory.clear();
+
+	Common::SeekableReadStream *dir = res.load("talk.lib");
+	dir->skip(4);		// Skip header
+
+	// Get the numer of entries
+	_directory.resize(dir->readUint16LE());
+
+	// Read in each entry
+	char buffer[17];
+	for (uint idx = 0; idx < _directory.size(); ++idx) {
+		dir->read(buffer, 17);
+		buffer[16] = '\0';
+
+		_directory[idx] = Common::String(buffer);
+	}
+
+	delete dir;
+
+	// Load in the locations stored in journal.txt
+	Common::SeekableReadStream *loc = res.load("journal.txt");
+
+	_locations.clear();
+	while (loc->pos() < loc->size()) {
+		Common::String line;
+		char c;
+		while ((c = loc->readByte()) != 0)
+			line += c;
+
+		_locations.push_back(line);
+	}
+
+	delete loc;
+}
+
+/**
+ * Loads the description for the current display index in the journal, and then
+ * word wraps the result to prepare it for being displayed
+ */
+int Journal::loadJournalFile(bool alreadyLoaded) {
```

Comment:
This is a very confusing function.

I think it should be changed to return \_lines instead of \_lines.size(), and get an extra argument for the index of the journal entry to be loaded.

It will then be much more suitable for the ways in which it is called.